src/ja/server/scheduler/scheduler.py

```diff
@@ -11,20 +16,64 @@ class Scheduler:
     jobs which are still runnable.
     """
 
-    def __init__(self, algorithm: SchedulingAlgorithm):
+    def __init__(self, algorithm: SchedulingAlgorithm, dispatcher: Dispatcher, special_resources: Dict[str, int]):
         """!
         Initialize a Scheduler.
 
-        @param algorithm The scheduling algorithm to use to assign jobs to
-          work machines.
+        @param algorithm The scheduling algorithm to use to assign jobs to work machines.
+        @param dispatcher The dispatcher to use for sending commands to the work machines.
+        @param special_resources The available special resources.
         """
+        self._algorithm = algorithm
+        self._dispatcher = dispatcher
+        self._special_resources = deepcopy(special_resources)
```

Comment:
Why is there the need to use deepcopy.

Reply:
Even if there is no specific need it is good practice I think.
Otherwise you might get strange bugs because two objects are changing the same list.

Reply:
On line 74 the special resources are being manipulated.
In this case you should definitely make a deepcopy I think.